File: src-tauri/src/app/handlers/db/musicians.rs
use tauri::{command, AppHandle};

use futures::executor::block_on;

use crate::state::ServiceAccess;

use crate::services::musicians;

#[command]
pub fn musicians_get_all(app_handle: AppHandle) -> Result<Vec<serde_json::Value>, String> {
    let result = app_handle.db(|db| block_on(musicians::get_all(db)));
    match result {
        Ok(musicians) => Ok(musicians),
        Err(e) => Err(e.to_string()),
    }
}

#[command]
pub fn musicians_get_by_id(app_handle: AppHandle, id: i32) -> Result<serde_json::Value, String> {
    let result = app_handle.db(|db| block_on(musicians::get_by_id(db, id)));
    match result {
        Ok(musician) => Ok(musician),
        Err(e) => Err(e.to_string()),
    }
}

#[command]
pub fn musicians_add(
    app_handle: AppHandle,
    first_name: String,
    last_name: Option<String>,
) -> Result<i32, String> {
    let result = app_handle.db(|db| block_on(musicians::add(db, first_name, last_name)));
    match result {
        Ok(id) => Ok(id),
        Err(e) => Err(e.to_string()),
    }
}

#[command]
pub fn musicians_update(
    app_handle: AppHandle,
    id: i32,
    first_name: String,
    last_name: Option<String>,
) -> Result<(), String> {
    let result = app_handle.db(|db| block_on(musicians::update(db, id, first_name, last_name)));
    match result {
        Ok(_) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}

#[command]
pub fn musicians_delete(app_handle: AppHandle, id: i32) -> Result<(), String> {
    let result = app_handle.db(|db| block_on(musicians::delete(db, id)));
    match result {
        Ok(_) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}

File: src/invokers/db/musicians.ts
import { Musician } from "@/types";
import { invoke } from "@tauri-apps/api";

export async function musiciansGetAll() {
  const musicians: Musician[] = await invoke("musicians_get_all");

  return musicians;
}

export async function musiciansDelete({ id }: { id: number }) {
  await invoke("musicians_delete", {
    id,
  });
}

export async function musiciansAdd({
  firstName,
  lastName,
}: {
  firstName: string;
  lastName?: string;
}) {
  const musicianId: number = await invoke("musicians_add", {
    firstName,
    lastName,
  });

  return musicianId;
}

export async function musiciansUpdate({
  id,
  firstName,
  lastName,
}: {
  id: number;
  firstName: string;
  lastName?: string;
}) {
  await invoke("musicians_update", {
    id,
    firstName,
    lastName,
  });
}

export async function musiciansGet({ id }: { id: number }) {
  const musician: Musician = await invoke("musicians_get_by_id", {
    id,
  });

  return musician;
}
